### PR TITLE
Update to react 16 react-native 49

### PR DESCRIPTION
**Pull request by paweldomas**
Depends on https://github.com/jitsi/jitsi-meet/pull/2105

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -53,10 +53,10 @@
     "nuclear-js": "1.4.0",
     "postis": "2.2.0",
     "prop-types": "15.6.0",
-    "react": "15.6.1",
-    "react-dom": "15.6.1",
+    "react": "16.0.0",
+    "react-dom": "16.0.0",
     "react-i18next": "4.8.0",
-    "react-native": "0.42.3",
+    "react-native": "0.49.3",
```
Comment (saghul) [inline package.json]: can we pin this to a commit which includes the fix for the FIXME above instead?

File: react/features/app/functions.web.js [hunk 2]
```diff
@@ -39,7 +39,8 @@ const _INTERCEPT_COMPONENT_RULES = [
      * @returns {UnsupportedMobileBrowser|void} If the rule is satisfied then
      * we should intercept existing component by UnsupportedMobileBrowser.
      */
-    () => {
+    // eslint-disable-next-line no-unused-vars
+    state => {
```
Comment (saghul) [inline react/features/app/functions.web.js]: why is this change needed? does flow complain?

**Comment by saghul** [comment]
Test results:

### iOS: worky worky with some problems:

- Looks like RCTBatchedBridge is now deprecated, I guess we should address that? https://fb.me/react-cxx-bridge
- Rotation seems broken, cannot repro on the current release so it might be related to the RN upgrade:

img_0005

<hr/>

### Android: no worky worky

screenshot_20171030-110412

File: react/features/app/functions.web.js (hunk 2 above)
Comment (paweldomas) [reply react/features/app/functions.web.js]: yes, flow first complains about incompatible type of the function and then eslint complains about unused vars. We were trying few approaches to fix and this one was the simplest.

File: package.json (hunk 1 above)
Comment (paweldomas) [reply package.json]: which FIXME ?

Comment (saghul) [reply package.json]: https://github.com/jitsi/jitsi-meet/pull/2105/files#diff-a9d0201e8c5bdc946e54218bb9206174R26

File: react/features/app/functions.web.js (hunk 2 above)
Comment (saghul) [reply react/features/app/functions.web.js]: Ack!

File: package.json (hunk 1 above)
Comment (paweldomas) [reply package.json]: do you know how to pin to a commit ? (I will google it otherwise, but just thought you may now)

Comment (paweldomas) [reply package.json]: this commit is contained in RN v0.50.0-rc.2 but then we'd be on not stable. I'm not sure if it's worth the effort - it's just a flow ignore

**Comment by paweldomas** [comment]
About the Android not worky worky: it turns out that it works with "remote JS debugger enabled", but doesn't work without it and there's this error about invariants violation

File: package.json (hunk 1 above)
Comment (saghul) [reply package.json]: Ack

**Comment by paweldomas** [comment]
Hey @saghul can you do a final review ? I think I've fixed all the issues.

**Comment by paweldomas** [comment]
About the batched bridge deprecation we were thinking with @lyubomir to not do that now since it's not a requirement. If you think it's super important then we can schedule it for a follow up PR.

File: react/index.native.js [hunk 3]
```diff
@@ -1,3 +1,27 @@
...
+// As an alternative solution we could stop using/polyfilling Symbols and
+// replace with classpath string constants or some kind of a wrapper around
+// that.
+import 'react-native/Libraries/ReactNative/renderApplication';
+
+// FIXME: remove once react-native-webrtc and react-native-prompt imports
+// PropTypes from the 'prop-types' lib instead of 'react'.
+import './features/base/react/prop-types-polyfill.js';
+
 import 'es6-symbol/implement';
```
Comment (lyubomir) [inline react/index.native.js]: @paweldomas, could you please clarify the following:

1. Today you told me that react-native on iOS now has `Symbol`. Which means that `import 'es6-symbol/implement';` is not necessary on iOS?

2. If the answer to 1. is positive, does it mean that `import 'react-native/Libraries/ReactNative/renderApplication';` is only necessary on Android then?

3. If the answer to 2. is positive, then why don't we have `import 'react-native/Libraries/ReactNative/renderApplication';` and `import 'es6-symbol/implement';` in index.android.js where there's practically no source code at the moment?

File: react/features/base/redux/functions.js [hunk 4]
```diff
@@ -19,7 +19,7 @@ export function assign(target: Object, source: Object) {
     let t = target;
 
     for (const property in source) { // eslint-disable-line guard-for-in
-        t = set(t, property, source[property], t === target);
+        t = set(t, property, source[property]);
```
Comment (lyubomir) [inline react/features/base/redux/functions.js]: @paweldomas & @saghul, are you sure that's the correct fix? I think I meant to use `_set` here. The intention of the code is (1) to return `target` if none of the setters lead to an actual change, (2) to not modify `target`, and (3) to do it with a 1 `copyOnWrite` at most.

File: react/features/mobile/external-api/middleware.js [hunk 5]
```diff
@@ -83,7 +83,7 @@ function _toErrorString(
         error
             ? typeof error === 'string'
                 ? error
-                : Error.prototype.toString(error)
+                : Error.prototype.toString.apply(error)
```
Comment (lyubomir) [inline react/features/mobile/external-api/middleware.js]: @paweldomas and/or @saghul, please take this commit out of this PR and push it into master!

File: react/features/base/redux/functions.js (hunk 4 above)
Comment (saghul) [reply react/features/base/redux/functions.js]: @lyubomir That sounds about right. I'll make the change and push that commit separately.

File: react/index.native.js (hunk 3 above)
Comment (saghul) [reply react/index.native.js]: @lyubomir That seems correct. iOS >= 9 has Symbol, and that's what we currently support. So it's a problem on Android only AFAICT. I'll be testing this today.

**Comment by saghul** [comment]
@lyubomir Moved the imports to index.android.js, where we need them. https://github.com/jitsi/jitsi-meet/pull/2105/commits/a78f6c205d9a1e6dcd246b5eb3b804469b785ca1 I'll do some testing (basic stuff) and merge if all goes well.

**Comment by saghul** [comment]
Jenkins test this please.
